### ISSUE TITLE
[no-type-alias] Cannot express complex types with generics

### ISSUE THREAD

**Issue by MichalLytek** [typescript-eslint/typescript-eslint#1317]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-type-alias": [
      "error",
      {
        "allowAliases": "in-unions-and-intersections",
        "allowCallbacks": "always",
        "allowLiterals": "always",
        "allowMappedTypes": "always",
        "allowConstructors": "always"
      }
    ]
  }
}
```

```TS
type RecursiveArray<TValue> = Array<RecursiveArray<TValue> | TValue>;
```

**Expected Result**

Allow to construct complex types using generics, the same as it allows using mapped types.

**Actual Result**

> Type aliases are not allowed. eslint(@typescript-eslint/no-type-alias)

**Additional Info**

This type is useful to model requirements on data like `["foo"]`, `[[["foo"]]]`, etc.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.10.0` |
| `@typescript-eslint/parser`        | `2.10.0` |
| `TypeScript`                       | `3.7.3` |
| `ESLint`                           | `6.7.2` |
| `node`                             | `12.13.0` |
| `npm`                              | `6.12.0` |

**Comment by bradzacher** [comment]
Unless I'm mistaken, this example is not actually complex, and is expressable via an interface:

```ts
interface RecursiveArray<TValue> extends Array<RecursiveArray<TValue> | TValue> { }
```

**Comment by MichalLytek** [comment]
> is expressable via an interface

This conflicts with another rule:
`An interface declaring no members is equivalent to its supertype.eslint(@typescript-eslint/no-empty-interface)`
So maybe I should add an issue about this rule for generics?

Also, the real world usage is to model the single nested tuple:
`type RecursiveSingleTuple<TValue> = [RecursiveSingleTuple<TValue> | TValue];`

So as we see, it should be possible to allow this generic types operation, as for many users `no-type-alias` is mostly for forbidding `type Student = User` or other nonsense types.

**Comment by bradzacher** [comment]
Hmm, you've run across this, because of your config:
- you use the generic form of arrays (`array-type`)
- you blacklist empty interfaces (`no-empty-interface`)
- you blacklist alias types (`no-type-alias`)

Workaround is to just eslint-disable for the one line.
It's a pretty rare case, no doubt.

Detecting recursive types is cumbersome, as you can nest them infinitely deep.
I'll leave this open for now. Happy to accept a PR.
Not sure what the best solution is.

**Comment by armano2** [comment]
This issue has been fixed [playground](https://typescript-eslint.io/play/#ts=4.9.5&sourceType=module&code=C4TwDgpgBAShDGBXATgZwJYDcIEFnIEMQAeAFQDUCAbRCAPigF4o9CS4k0td8izKa9KAB8oFarToBuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1qNNoENGfVBgDa4bDkTRoHaJAA04iVgnZIA+BwDuAQUHIU6KM1qwmlDk2T8mAE1rN8Ug2XwWrCpavXxNWgMIaAEZ8ZADWwlACWnyEqIqqahraADKUTtACkT4xcZ6JOMlaALJ8xKS2ACokhhg5sfFeEj5+-pbI+HCustnRDfkFLdoA4ohMUpRkvfC5qE1gAL5eALriSwtAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) see #3865 

note: `allowGenerics` has to be enabled
